### ISSUE TITLE
Type cannot be assigned to type "None"

### ISSUE THREAD

**Issue by jgil-r** [microsoft/pyright#3521]
**Describe the bug**

Not sure if this is expected behavior

```
    Cannot assign member "next" for type "Node"
    Expression of type "Node" cannot be assigned to member "next" of class "Node"
    Type cannot be assigned to type "Node"
```

**To Reproduce**
Python 3.10.4
pyright 1.1.249

Using neovim lsp, but if I run `pyright main.py` which holds the code below I also get the following output:

```
No configuration file found.
No pyproject.toml file found.
stubPath /home/jg/projects/pydsa/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/jg/projects/pydsa/main.py
  /home/jg/projects/pydsa/main.py:25:23 - error: Cannot assign member "next" for type "Node"
    Expression of type "Node" cannot be assigned to member "next" of class "Node"
      Type cannot be assigned to type "None" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.802sec
```

```python
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self, value):
        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.length = 1

    def print_list(self):
        temp = self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next

    def append(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node    #    <-- error is here
            self.tail = new_node
        self.length += 1
        return True

my_linked_list = LinkedList(1)
my_linked_list.append(2)
my_linked_list.append(3)

my_linked_list.print_list()
```

**Comment by erictraut** [comment]
The problem is that the class `Node` provides no type annotations for its instance variables `value` and `next`. Pyright therefore infers the types of these values based on the values they are assigned. In the case of `next`, the inferred type is `Unknown` because it is assigned `value`, and `value` is an input parameter with no declared type. In the case of `next`, the inferred type is `None` because that is the only value assigned to it within the `Node` class.

If you add the following type annotations to your implementation of `Node`, the error will go away:
```python
class Node:
    def __init__(self, value: "Node"):
        self.value = value
        self.next: Node | None = None
```

**Comment by jgil-r** [comment]
That makes sense. I appreciate you taking the time to explain!
